readme.md

# usb-cdc-acm

Userspace javascript implementation of a USB CDC ACM driver, on top of libusb.

[![Build Status](https://travis-ci.org/NordicPlayground/node-usb-cdc-acm.svg?branch=master)](https://travis-ci.org/NordicPlayground/node-usb-cdc-acm)

This is part of [Nordic Semiconductor](http://www.nordicsemi.com/)'s javascript tools to
interface with nRF SoCs and development kits. Although USB CDC ACM is part of the USB specifications
and a *de facto* standard when it comes to emulating a serial port connection on an embedded device, 
Nordic Semiconductor cannot offer support for other hardware platforms. This software is provided "as is".

## Motivation

Sometimes you want to fetch info from the USB descriptors via the 
[NodeJS `usb` module](https://github.com/tessel/node-usb), and at the same time use the CDC ACM 
interface to send and receive data. But this is not possible with some host configurations
(notably, win32/win64 platforms and their need to manually switch the driver via Zadig or the like).

## API

`usb-cdc-acm` provides a [duplex `Stream` interface](https://nodejs.org/api/stream.html). Please
refer to NodeJS's documentation about `Stream`s for a full API spec.

Other than that, `usb-cdc-acm` provides two entry points. Either of those need you to use `usb` to
fetch a reference to the `Device` or to a CDC `Interface`.

Use the factory method when your USB device only has one CDC ACM interface:
```js
var usb = require('usb');
const UsbCdcAcm = require('usb-cdc-acm');

var device = usb.findByIds( 0x1915, 0x520f ); // VID/PID for Nordic Semi / USB CDC demo

// The device MUST be open before instantiating the UsbCdcAcm stream!
device.open();

// An options object with the baud rate is optional.
let stream = UsbCdcAcm.fromUsbDevice(device, { baudRate: 1000000});

// Then, use it as any other Stream
stream.on('data', function(data) { console.log('recv: ', data); });
stream.write('Hello world!');

// Remember to destroy the stream and close the device when finished!
// Failure to do so might leave the USB device in an ususable state for other applications.
setTimeout(function(){ 
    stream.destroy();
    device.close();
}, 5000);
```

If the device has several CDC ACM interfaces and finer control is needed, 
```js
// An options object with the baud rate is optional.
let stream = new UsbCdcAcm(device.interfaces[0], { baudRate: 1000000});
```

For a more complete example, check the `test/test.js` file.


## Legal

Distributed under a BSD-3 license. See the `LICENSE` file for details.



CHANGELOG

# v0.1.1 (2018-01-24)

* Worked around calling `isKernelDriverAttached()` in win32 platforms, where it's unsupported.

# v0.1.0 (2018-01-23)

* Initial release


test/test.js

/**
 * 
 * This demo shows how to get an instance of an usb-cdc-acm Stream,
 * and send/receive data from it.
 * 
 * It's quite important to both destroy the stream and close the device
 * when done.
 * 
 */

const usb = require('usb');
const UsbCdcAcm = require('..');

const Debug = require('debug');
const debug = Debug('main');

Debug.enable('*');

// let device = usb.findByIds( 0x1915, 0x521f );   // VID/PID for Nordic semi / USB SDFU
let device = usb.findByIds( 0x1915, 0x520f );   // VID/PID for Nordic semi / USB CDC demo
// let device = usb.findByIds( 0x1366, 0x1015 );   // VID/PID for a Segger IMCU (with USB storage)
// let device = usb.findByIds( 0x1366, 0x0105 );   // VID/PID for a Segger IMCU (without USB storage)

// console.log(device);

device.timeout = 100;
debug('Opening device');
device.open();

let ifaces = device.interfaces;

// usb.setDebugLevel(4); // Uncomment for extra USB verbosiness

// let stream = UsbCdcAcm.fromUsbDevice(device, { baudRate: 115200 });
let stream = UsbCdcAcm.fromUsbDevice(device, { baudRate: 1000000});

// Display all data received
stream.on('data', (data)=>{ debug('data', data.toString()); });

// Log other events from the Stream, just in case
stream.on('error', (err)=>{ debug('error',err); });
stream.on('status', (sts)=>{ debug('status', sts); });
stream.on('close', ()=>{debug('Stream is now closed')});
stream.on('drain', ()=>{debug('Stream can be drained now')});

let i = 0

let timer = setInterval(()=>{
    const written = stream.write(`foobar ${i++} ${Date()}\n`);
    debug('Sent a write');
}, 2500);


setTimeout(()=>{
    clearInterval(timer);
    
    debug('Closing the stream');
    stream.destroy();
    
    setTimeout(()=>{
        debug('Closing device');
//         device.close();
    }, 5000);
    
}, 50000);



src/usb-stream.js

// import EventEmitter from 'events';
// import { Duplex } from 'stream';
// import Debug from 'debug';
// import usb from 'usb';
const Duplex = require('stream').Duplex;
const Debug = require('debug');
const usb = require('usb');

const debug = Debug('usb-stream');


// Utility function.
// Given an interface, assert that it's valid for instantiating a stream from it
// Specifically, the interface must have only one 
// "in" bulk endpoint and one "out" bulk endpoint.
function assertInterface(iface) {
    const endpoints = iface.endpoints;

    return (
        // Only two endpoints, and 
        endpoints.length === 2 && 
        // both are bulk transfer, and
        endpoints[0].transferType === usb.LIBUSB_TRANSFER_TYPE_BULK &&
        endpoints[1].transferType === usb.LIBUSB_TRANSFER_TYPE_BULK &&
        // their direction (in/out) is different
        endpoints[0].direction !== endpoints[1].direction
    );
}


// export class UsbStream extends Duplex {
class UsbStream extends Duplex {
    constructor(iface) {
        
        if (!assertInterface(iface) ){
            throw new Error('Interface is not valid (does not have one "in" bulk endpoint and one "out" bulk endpoint)');
        }
        
        super();
        
        this.iface = iface;
        this.device = iface.device;
        
        const endpoints = iface.endpoints;

        if (endpoints[0].direction === 'in') {
            this.in = endpoints[0];
            this.out = endpoints[1];
        } else {
            this.in = endpoints[1];
            this.out = endpoints[0];
        }
        
        if (iface.isKernelDriverActive()) {
            iface.detachKernelDriver();
        }
        iface.claim();
        
        // Perform a SET_LINE_CODING request
        
        // Perform a USB_CDC_REQ_SET_CONTROL_LINE_STATE (0x22) control transfer
        // This is documented in the PSTN doc of the USB spec, section 6.3.12
        this._controlTransfer(
            0x21, // bmRequestType: [host-to-device, type: class, recipient: iface]
            0x22, // SET_CONTROL_LINE_STATE
            0x03, // 0x02 "Activate carrier" & 0x01 "DTE is present"
            0x00, // index 0 = interface index ???
            new Buffer([])     // No data expected back
//         ).then(()=>{
//             return this._controlTransfer(
//                 0x21, // bmRequestType: [host-to-device, type: class, recipient: iface]
//                 0x20, // SET_LINE_CODING
//                 0x01, // value 0x0
//                 0x00, // index 0
//                 new Uint8Array([0x80, 0x25, 0, 0, 0, 0, 0x08])
//         )
//         })
        ).then( ()=>{
            debug('Control transfer 0x21 0x20 performed');
            
            this.in.on('data', this._onData.bind(this));
            this.in.on('error', (err)=>this.emit('error', err));
            this.out.on('error', (err)=>this.emit('error', err));
            
            this.in.timeout = 1000;
            this.out.timeout = 1000;
        });
        
        
        debug(this.descriptor);
        
//         if (!desc || desc.
    }
    
    _read(){
        debug('_read');
        if (!this.polling) {
            debug('starting polling');
            this.in.startPoll();
            this.polling = true;
        }
    }
    
    _onData(data) {
        debug('_onData ' + data);
        const keepReading = this.push(data);
        if (!keepReading) {
            this._stopPolling();
        }
    }
    
    _onError(err) {
        debug('Error: ', err)
        throw err;
    }
    
    _stopPolling() {
        debug('_stopPolling');
        if (this.polling) {
            debug('stopping polling');
            this.in.stopPoll();
            this.polling = false;
        }
    }
    
    _write(data){
        debug('_write ' + data.toString());

        this.out.transfer(data, (err)=>{
            if (err) {
                debug('Out transfer error: ', err);
                this.emit(err)
            }
        });
    }
    
//     _writev(){}
    
    _final(){
        debug('_final');
        this._stopPolling();
        iface.release();
        iface.attachKernelDriver();
    }
    
    // The device's controlTransfer, wrapped as a Promise
    _controlTransfer(bmRequestType, bRequest, wValue, wIndex, data_or_length) {
        return Promise.resolve();
//         return new Promise((res, rej)=>{
//             this.device.controlTransfer(
//                 bmRequestType, 
//                 bRequest, 
//                 wValue, 
//                 wIndex, 
//                 data_or_length, 
//                 (err, data)=>{ err ? rej(err) : res(data); })
//         });
    }
    
    // Given an instance of Device (from the 'usb' library), opens it, looks through
    // its interfaces, and creates an instance of UsbStream on the first interface which
    // looks like a serial port (having only one "in" bulk endpoint and one 
    // "out" bulk endpoint, see assertInterface() ). 
    // 
    // The given Device must be already open()ed. Conversely, it has to be close()d
    // when the stream is no longer used, or if this method throws an error.
    //
    // Returns an instance of UsbStream.
    static fromUsbDevice(device){
        let ifaces = device.interfaces;
        
        for(let i=0,l=ifaces.length; i<l; i++) {
            const iface = ifaces[i];
            const endpoints = ifaces[i].endpoints;
            
//             debug(iface.descriptor);
//             debug(iface.endpoints);
//             debug(iface.endpoints.map(e=>e.direction));
//             debug(endpoints);
            
            if (assertInterface(iface)) {
               return new UsbStream(iface); 
            }
        }
        
        throw new Error('No valid interfaces found in USB device (they do not have one "in" bulk endpoint and one "out" bulk endpoint)');
    }

}




module.exports = UsbStream;


src/split-descriptors.js
// Quasi-trivial utility to parse USB descriptors from a Uint8Array
// The first byte in the descriptor is the descriptor length, and they are just
// concatenated together, so something like:
// 5 X X X X 4 X X X 9 X X X X X X X X
// should be splitted into
// 5 X X X X  |  4 X X X   |  9 X X X X X X X X


// Given a Uint8Array, returns an Array of Uint8Array
// Each element of the resulting array is a subarray of the original Uint8Array.
export default function splitDescriptors(bytes) {
    let len = bytes.length;
    const descs = [];
    let pointer = 0;

    while (len > 0) {
        const descLen = bytes[pointer];
        descs.push(bytes.subarray(pointer, pointer + descLen));
        len -= descLen;
        pointer += descLen;
    }

    return descs;
}


/*
let test = Uint8Array.from([5, 36, 0, 16, 1, 5, 36, 1, 3, 1, 4, 36, 2, 6, 5, 36, 6, 0, 1]);

console.log(splitDescriptors(test));
*/

/*
The previous code should output:

[ Uint8Array [ 5, 36, 0, 16, 1 ],
  Uint8Array [ 5, 36, 1, 3, 1 ],
  Uint8Array [ 4, 36, 2, 6 ],
  Uint8Array [ 5, 36, 6, 0, 1 ] ]
*/



src/usb-cdc-acm.js

import { Duplex } from 'stream';
import Debug from 'debug';
import usb from 'usb';
import splitDescriptors from './split-descriptors';
// const Duplex = require('stream').Duplex;
// const Debug = require('debug');
// const usb = require('usb');
// const util = require('util');

// Two debug levels: one for initialization/teardown messages, and one
// for logging all data being sent/recv around
const debugInfo = Debug('usb-cdc-acm:info');
const debugData = Debug('usb-cdc-acm:data');

// const splitDescriptors = require('./split-descriptors');


// Utility function.
// Given an interface, assert that it looks like a CDC management interface
// Specifically, the interface must have only one
// "out" interrupt endpoint, and a CDC Union descriptor.
// Will return boolean `false` if the interface is not valid,
// or an integer number (corresponding to the associated data interface)
function assertCdcInterface(iface) {
    const { endpoints, descriptor } = iface;

    if (descriptor.bInterfaceClass !== usb.LIBUSB_CLASS_COMM || // 2, CDC
        descriptor.bInterfaceSubClass !== 2) { // ACM
        return false;
    }

    // Check it has only one endpoint, and of the right kind
    if (endpoints.length !== 1 ||
        endpoints[0].transferType !== usb.LIBUSB_TRANSFER_TYPE_INTERRUPT ||
        endpoints[0].direction !== 'in') {
        return false;
    }

    // node-usb doesn't parse the CDC Union descriptor inside the interface
    // descriptor, so parse and find it manually here.
    const additionalDescriptors = splitDescriptors(descriptor.extra);
    let slaveInterfaceId = false;

    for (let i = 0, l = additionalDescriptors.length; i < l; i += 1) {
        const desc = additionalDescriptors[i];

        // 0x24 = class-specific descriptor. 0x06 = CDC Union descriptor
        if (desc[1] === 0x24 && desc[2] === 6) {
            if (desc[3] !== iface.id) {
                // Master interface should be the current one!!
                return false;
            }
            [,,,, slaveInterfaceId] = desc; // slaveInterfaceId = desc[4];
        }
    }

    if (slaveInterfaceId === false) {
        // CDC Union descriptor not found, this is not a well-formed USB CDC ACM interface
        return false;
    }

    return (slaveInterfaceId);
}


// Utility function.
// Given an interface, assert that it looks like a CDC data interface
// Specifically, the interface must have only one
// "in" bulk endpoint and one "out" bulk endpoint.
function assertDataInterface(iface) {
    const { endpoints } = iface;

    return (
        // Right class (0x0A)
        iface.descriptor.bInterfaceClass === usb.LIBUSB_CLASS_DATA &&
        // Only two endpoints, and
        endpoints.length === 2 &&
        // both are bulk transfer, and
        endpoints[0].transferType === usb.LIBUSB_TRANSFER_TYPE_BULK &&
        endpoints[1].transferType === usb.LIBUSB_TRANSFER_TYPE_BULK &&
        // their direction (in/out) is different
        endpoints[0].direction !== endpoints[1].direction
    );
}


export default class UsbCdcAcm extends Duplex {
    constructor(ifaceCdc, options = {}) {
        const ifaceDataId = assertCdcInterface(ifaceCdc);
        if (ifaceDataId === false) {
            throw new Error('CDC interface is not valid');
        }

        const ifaceData = ifaceCdc.device.interfaces[ifaceDataId];
        if (!assertDataInterface(ifaceData)) {
            throw new Error('Data interface is not valid');
        }

        super(options);

        this.ifaceCdc = ifaceCdc;
        this.ifaceData = ifaceData;
        this.device = ifaceCdc.device;

        [this.ctr] = ifaceCdc.endpoints;

        if (ifaceData.endpoints[0].direction === 'in') {
            [this.in, this.out] = ifaceData.endpoints;
        } else {
            [this.out, this.in] = ifaceData.endpoints;
        }

        debugInfo('claiming interfaces');

        this._reattachCdcDriverAtFinal = false;
        this._reattachDataDriverAtFinal = false;
        // Linux/mac need to detach the cdc-acm kernel driver, but
        // windows users did that manually, and libusb-win just throws
        // errors when detaching/attaching kernel drivers.
        if (process.platform !== 'win32') {
            if (ifaceCdc.isKernelDriverActive()) {
                ifaceCdc.detachKernelDriver();
                this._reattachCdcDriverAtFinal = true;
            }

            if (ifaceData.isKernelDriverActive()) {
                ifaceData.detachKernelDriver();
                this._reattachDataDriverAtFinal = true;
            }
        }
        ifaceCdc.claim();
        ifaceData.claim();

        this.ctr.on('data', this._onStatus.bind(this));
        this.ctr.on('error', this._onError.bind(this));
        this.ctr.startPoll();


        // Set baud rate and serial line params,
        // then set the line as active
        this._controlSetLineCoding(options.baudRate || 9600)
            .then(() => { this._controlLineState(true); })
            .then(() => { this._controlGetLineCoding(); })
            .then(() => {
                this.in.on('data', data => this._onData(data));
                this.in.on('error', err => this.emit('error', err));
                this.out.on('error', err => this.emit('error', err));

                this.in.timeout = 1000;
                this.out.timeout = 1000;
            });
    }

    _read() {
        debugData('_read');
        if (!this.polling) {
            debugInfo('starting polling');
            this.in.startPoll();
            this.polling = true;
        }
    }

    _onData(data) {
        debugData('_onData ', data);
        const keepReading = this.push(data);
        if (!keepReading) {
            this._stopPolling();
        }
    }

    _onError(err) {
        debugInfo('Error: ', err);
        this.emit('error', err);
        //         throw err;
    }

    _onStatus(sts) { // eslint-disable-line class-methods-use-this
        debugInfo('Status: ', sts);
    }

    _stopPolling() {
        debugInfo('_stopPolling');
        if (this.polling) {
            debugInfo('stopping polling');
            this.in.stopPoll();
            this.polling = false;
        }
    }

    _write(data, encoding, callback) {
        debugData(`_write ${data.toString()}`);

        this.out.transfer(data, callback);
    }

    _destroy() {
        debugInfo('_destroy');

        // Set line state as unused, close all resources, release interfaces
        // (waiting until they are released), reattach kernel drivers if they
        // were attached before, then emit a 'close' event.

        this._controlLineState(false)
            .then(() => {
                this._stopPolling();
                this.ctr.stopPoll();

                this.ctr.removeAllListeners();
                this.in.removeAllListeners();
                this.out.removeAllListeners();

                this.ifaceCdc.release(true, err => {
                    if (err) { throw err; }
                    this.ifaceData.release(true, err2 => {
                        if (err2) { throw err2; }

                        if (this._reattachCdcDriverAtFinal) {
                            this.ifaceCdc.attachKernelDriver();
                        }
                        if (this._reattachDataDriverAtFinal) {
                            this.ifaceData.attachKernelDriver();
                        }

                        debugInfo('All resources released');
                        this.emit('close');
                    });
                });
            });
    }


    // Performs a _controlTransfer() to set the line state.
    // Set active to a truthy value to indicate there is something connected to the line,
    // falsy otherwise.
    // Returns a Promise.
    _controlLineState(active) {
        // This is documented in the PSTN doc of the USB spec, section 6.3.12
        return this._controlTransfer(
            0x21, // bmRequestType: [host-to-device, type: class, recipient: iface]
            0x22, // SET_CONTROL_LINE_STATE
            active ? 0x03 : 0x00, // 0x02 "Activate carrier" & 0x01 "DTE is present"
            this.ifaceCdc.id, // interface index
            Buffer.from([]) // No data expected back
        );
    }

    // Performs a _controlTransfer to set the line coding.
    // This includes bitrate, stop bits, parity, and data bits.
    _controlSetLineCoding(baudRate = 9600) {
        // This is documented in the PSTN doc of the USB spec, section 6.3.10,
        // values for the data structure at the table in 6.3.11.
        const data = Buffer.from([
            0, 0, 0, 0, // Four bytes for the bitrate, will be filled in later.
            0, // Stop bits. 0 means "1 stop bit"
            0, // Parity. 0 means "no parity"
            8, // Number of data bits
        ]);

        data.writeInt32LE(baudRate, 0);

        debugInfo('Setting baud rate to ', baudRate);

        return this._controlTransfer(
            0x21, // bmRequestType: [host-to-device, type: class, recipient: iface]
            0x20, // SET_LINE_CODING
            0x00, // Always zero
            this.ifaceCdc.id, // interface index
            data
        );
    }

    // Performs a _controlTransfer to get the line coding.
    // This includes bitrate, stop bits, parity, and data bits.
    _controlGetLineCoding() {
        // This is documented in the PSTN doc of the USB spec, section 6.3.11,
        debugInfo('Requesting actual line coding values');

        return this._controlTransfer(
            0xA1, // bmRequestType: [device-to-host, type: class, recipient: iface]
            0x21, // GET_LINE_CODING
            0x00, // Always zero
            this.ifaceCdc.id, // interface index
            7 // Length of data expected back
        ).then(data => {
            const baudRate = data.readInt32LE(0);
            const rawStopBits = data.readInt8(4);
            const rawParity = data.readInt8(5);
            const dataBits = data.readInt8(6);

            let stopBits;
            let parity;
            switch (rawStopBits) {
                case 0: stopBits = 1; break;
                case 1: stopBits = 1.5; break;
                case 2: stopBits = 2; break;
                default: throw new Error('Invalid value for stop bits received (during a GET_LINE_CODING request)');
            }
            switch (rawParity) {
                case 0: parity = 'none'; break;
                case 1: parity = 'odd'; break;
                case 2: parity = 'even'; break;
                case 3: parity = 'mark'; break;
                case 4: parity = 'space'; break;
                default: throw new Error('Invalid value for parity received (during a GET_LINE_CODING request)');
            }

            debugInfo('Got line coding: ', data);
            debugInfo('Reported baud rate: ', baudRate);
            debugInfo('Reported stop bits: ', stopBits);
            debugInfo('Reported parity: ', parity);
            debugInfo('Reported data bits: ', dataBits);

            return data;
        });
    }

    // The device's controlTransfer, wrapped as a Promise
    _controlTransfer(bmRequestType, bRequest, wValue, wIndex, dataOrLength) {
        return new Promise((res, rej) => {
            this.device.controlTransfer(
                bmRequestType,
                bRequest,
                wValue,
                wIndex,
                dataOrLength,
                ((err, data) => (err ? rej(err) : res(data)))
            );
        });
    }


    // Given an instance of Device (from the 'usb' library), opens it, looks through
    // its interfaces, and creates an instance of UsbStream per interface which
    // looks like a CDC ACM control interface (having the right descriptor and endpoints).
    //
    // The given Device must be already open()ed. Conversely, it has to be close()d
    // when the stream is no longer used, or if this method throws an error.
    //
    // Returns an array of instances of UsbCdcAcm.
    static fromUsbDevice(device, options = {}) {
        const ifaces = device.interfaces;

        for (let i = 0, l = ifaces.length; i < l; i += 1) {
            const iface = ifaces[i];

            if (assertCdcInterface(iface) !== false) {
                return new UsbCdcAcm(iface, options);
            }
        }

        throw new Error('No valid CDC interfaces found in USB device');
    }
}
